Remove global spinlock for EH stacktrace

janvorli opened the pull request:
The global spinlock that was used to ensure that stack trace and the associated dynamic methods array were updated and read atomically. However, for the new EH, it has shown to cause a high contention in case many threads were handling exceptions at the same time.

This change replaces the two arrays by one object member in the exception class. It contains reference to either the `byte[]` of the stack trace (when there are no dynamic methods on the stack trace) or an `object[]` where the first element contains the stack trace `byte[]` reference and the following elements contain what used to be in the dynamic array. That allows atomic updates and reads of the stack trace and dynamic method keepalive references without a need of a lock.

It improves the performance of exceptions being handled on multiple threads at the same time 3 fold. For example, a testing app running on my Windows x64 machine with 10 CPU cores that throws an exception on 10 threads in parallel 1000000 times over 10 stack frames took 30s of each thread's time before and 9 seconds after the change.

The original code was quite convoluted, it was difficult to reason about and it had some races in it that were hidden behind the global lock. So I have decided to rewrite the whole thing from scratch.

The way it ensures that it is race free is that whenever it updates the exception stack trace and the one that's on the exception was created by a different thread, it creates a deep copy of both the stack trace and the keepalive array. When making the copy, it also handles a case when a frame that needs a keepalive entry is on the stack trace part, but the keepalive array extracted from the exception is stale (the other thread needed to resize the keepalive array, but not the stack trace). In that case, the stack trace is trimmed at first such entry found.

Since the case when multiple threads are throwing the same exception and so they are modifying its stack trace in parallel is pathological anyways, I believe the extra work spent on creating the clones of the arrays is a good tradeoff for ensuring easy to reason about thread safety.

I have also merged the `StackTraceInfo::AppendElement` and the `StackTraceInfo::SaveStackTrace` into one, as they were always being called in pair for a single frame. Before, the first one needed to store the frame and the second would extract it from that storage and append it to the stack trace.

Finally, since with the previous iteration of this change, a bug in building the stack trace was found, I have added a coreclr test to verify stack trace for an exception matches the expectations.

janvorli committed:
Remove global spinlock for EH stacktrace

The global spinlock that was used to ensure that stack trace and the
associated dynamic methods array were updated and read atomically.
However, for the new EH, it has shown to cause a high contention in case
many threads were handling exceptions at the same time.

This change replaces the two arrays by one object member in the
exception class. It contains reference to either the byte[] of the
stack trace (when there are no dynamic methods on the stack trace) or an
object[] where the first element contains the stack trace byte[]
reference and the following elements contain what used to be in the
dynamic array. That allows atomic updates and reads of the stack trace
and dynamic method keepalive references without a need of a lock.

The original code was quite convoluted, it was difficult to reason about
and it had some races in it that were hidden behind the global lock. So
I have decided to rewrite the whole thing from scratch.

The way it ensures that it is race free is that whenever it updates the
exception stack trace and the one that's on the exception was created by
a different thread, it creates a deep copy of both the stack trace and
the keepalive array. When making the copy, it also handles a case when
a frame that needs a keepalive entry is on the stack trace part, but the
keepalive array extracted from the exception is stale (the other thread
needed to resize the keepalive array, but not the stack trace). In that
case, the stack trace is trimmed at first such entry found.

Since the case when multiple threads are throwing the same exception and
so they are modifying its stack trace in parallel is pathological
anyways, I believe the extra work spent on creating the clones of the arrays
is a good tradeoff for ensuring easy to reason about thread safety.

I have also removed a dead code path from the
StackTraceInfo::SaveStackTrace.

Finally, since with the previous iteration of this change, a bug in
building the stack trace was found, I have added a coreclr test to
verify stack trace for an exception matches the expectations.

janvorli committed:
Fix MUSL build

janvorli committed:
Fix x86 build

janvorli commented:
There are some test errors, I am investigating them.

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3226,196 +2858,326 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
+}
+#endif // !TARGET_UNIX
+
+// Ensure that there is space for one more element in the stack trace array.
```
```suggestion
// Ensure that there is space for neededSize elements in the stack trace array.
```

janvorli replied:
Changed

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3226,196 +2858,326 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
+            STRESS_LOG3(LF_EH, LL_INFO1000, "EnsureStackTraceArray copying from stackTrace=%p to stackTrace=%p, owner thread was %04x\n", OBJECTREFToObject(pStackTrace->Get()), OBJECTREFToObject(newStackTrace.Get()), pStackTrace->GetObjectThread()->GetOSThreadId());
+            // Copy the original array to the new one
+            newStackTrace.CopyDataFrom(*pStackTrace);
+            // TODO: cleanup this - make the method have delta as the parameter instead
```
Fix TODO?

janvorli replied:
I think that that line of code the comment refers to is stale from the time when I wasn't returning copies of the array from GetStackTrace for stack traces created by external threads. I'll double check.

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1473,62 +1473,38 @@ void StackTraceArray::CheckState() const
     CONTRACTL_END;
 
-    size_t raw_size = grow_size * sizeof(StackTraceElement) + sizeof(ArrayHeader);
+    size_t raw_size = size * sizeof(StackTraceElement) + sizeof(ArrayHeader);
```
This multiplication can overflow

janvorli replied:
Hmm, right, on 32 bit OSes. I haven't noticed there was this preexisting issue, thank you for noticing it.

janvorli replied:
Fixed

jkotas commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -87,111 +123,83 @@ FCIMPL3(VOID, ExceptionNative::GetStackTracesDeepCopy, Object* pExceptionObjectU
-    {
-        gc.dynamicMethodsArray = (PTRARRAYREF)ObjectToOBJECTREF(pDynamicMethodsUnsafe);
-    }
+    ExceptionObject::GetStackTraceParts(gc.refStackTrace, gc.stackTrace, &gc.refKeepaliveArray);
```
This does way too many copies. This will be often called by async on different thread than the exception was thrown originally that will cause `ExceptionObject::GetStackTraceParts` to create a copy. The `DeepCopyStackTrace` is then going to create second copy.

I think we can avoid all copies in this method by:
- Setting the owning thread in `ExceptionNative::GetStackTracesDeepCopy` to (Thread*)1. It will signal that the stack trace is immutable. Any thread that will want to append anything to it will have make its own copy first.
- Once we do that, I think this method can be deleted and restoring of the stacktrace can be replace by a simple assignment, without any copies.

janvorli replied:
I've implemented this suggestion.

jkotas commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -76,8 +76,44 @@ FCIMPL0(VOID, ExceptionNative::PrepareForForeignExceptionRaise)
+
 // Given an exception object, this method will extract the stacktrace and dynamic method array and set them up for return to the caller.
-FCIMPL3(VOID, ExceptionNative::GetStackTracesDeepCopy, Object* pExceptionObjectUnsafe, Object **pStackTraceUnsafe, Object **pDynamicMethodsUnsafe);
+FCIMPL2(VOID, ExceptionNative::GetStackTracesDeepCopy, Object* pExceptionObjectUnsafe, Object **pStackTraceUnsafe);
```
```suggestion
FCIMPL1(Object *, ExceptionNative::GetStackTracesDeepCopy, Object* pExceptionObjectUnsafe)
```
This can return the value now that there is just a single value to return

jkotas replied:
Updated the suggestion to `FreezeStackTrace` since the method does not actually perform a deep copy in typical case anymore.

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1545,7 +1521,23 @@ void StackTraceArray::CopyFrom(StackTraceArray const & src)
     CONTRACTL_END;
 
-    m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(src.Capacity()));
+    m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(src.Get()->GetNumComponents()));
```
We may want to only allocate enough space for the entries that are filled in. We should not need to clone the empty slots.

jkotas replied:
Or even better - if you take my other suggestion, we can freeze the stack by setting its thread to `(Thread*)1`. It would make the storing and restoring of ExceptionDispatchInfo allocation-free. The copy would be created lazily only once we need to append more frames to the stack trace after the restored ExceptionDispatchInfo is rethrown.

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1899,119 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+                }
+                if (keepaliveObject == NULL)
+                {
+                    // Trim the stack trace at a point where a dynamic or collectible method is found without a corresponding keepalive object.
```
How is this possible?

If I understand the code correctly, the thread safety should be guaranteed by writing the `size` field last and reading it first. Is that right? (Whatever it is, it would be nice to have it documented in a comment somewhere.)

I do not think it should be possible to get methods items without corresponding keep alive items here. If it was possible, I suspect there may be situations where the method is not kept alive and the `pMethod->IsLCGMethod()` check above can crash.

janvorli replied:
Here is the case when it happens:
* Thread A owns the stack trace and is adding a new entry that requires a keep alive object.
* Thread B wants to update the stack trace too, so it fetches the stack trace array S1 and keep alive array K1 from the exception
* Thread A finds the keep alive array is full, so it creates a larger clone of the previous one, let's call it K2
* Thread A adds the new entry to keep alive array K2
* Thread A adds the new entry to stack trace array S1
* Thread A writes K2 to the exception
* Now thread B creates clones of S1 and K1 (when it read it from the exception, the K2 was not there yet). But S1 already has the new entry added by thread A, which is not covered by K1. Since K1 won't change anymore, we solve this by trimming the S1 at the element where we've found a frame that needs keep alive object and was not covered by K1

However, I have realized you are right that in such case, calling IsLCGMethod on the MethodTable above could crash. For example, if the thread A's exception was collected together with its keep alive array K2 that can be the only thing holding the method alive and if that would happen before the thread B called the IsLCGMethod.
So I need to figure out some other way to handle the situation described above.

jkotas commented:
> Since the case when multiple threads are throwing the same exception and so they are modifying its stack trace in parallel is pathological anyways, I believe the extra work spent on creating the clones of the arrays is a good tradeoff for ensuring easy to reason about thread safety.

I agree that multiple threads throwing the same exception is a pathological case that can be slow. One exception being caught on one thread and rethrown on different threads using ExceptionDispatchInfo is not that uncommon in async code.

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3226,196 +2858,326 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
+            if (m_keepaliveItemsCount == -1)
+            {
+                // The m_keepaliveItemsCount was not initialized yet, so we need to calculate it.
+                m_keepaliveItemsCount = GetKeepaliveItemsCount(&gc.stackTrace);
```
Should we remember the keep alive item count in the stack trace header, so that we do not have to recompute it here?

janvorli replied:
Done

janvorli commented on src/coreclr/vm/clrex.h:
```diff
@@ -51,28 +51,16 @@ struct StackTraceElement
     }
 };
 
-// This struct is used by SOS in the diagnostic repo.
```
This comment was not correct, there was no dependency in SOS on this struct. The only thing it depends on is the StackTraceElement.

janvorli committed:
Fix several issues

* Missing calls to IsOverflow at few places
* Added a flag on StackTraceElement to indicate that the element needs a
  keepalive entry. It removes the need to call IsLCGMethod / Collectible
  check on the method table stored in the element and eliminates a
  possible problem with the method being collected in one place.
* Returned missing call to StackFrameInfo::Init to the x86 code path
* Removed obsolete comment and code line

janvorli committed:
Few changes based on feedback

* Add keep alive items count to the stack trace header.
* Implement the concept of frozen stack traces to eliminate copies in
  the ExceptionDispatchInfo storing / restoring exceptions.

janvorli committed:
Rename keepalive to keepAlive

janvorli committed:
Handle possible array size overflow

In the StackTraceArray::Allocate

janvorli commented on src/coreclr/vm/clrex.h:
```diff
@@ -51,28 +51,16 @@ struct StackTraceElement
-    unsigned            m_dFrameCount;      // current frame in stack trace
-    unsigned            m_cDynamicMethodItems; // number of items in the Dynamic Method array
-    unsigned            m_dCurrentDynamicIndex; // index of the next location where the resolver object will be stored
+    int m_keepaliveItemsCount = -1; // -1 indicates the count is not initialized yet
```
Updated to keepAlive everywhere

jkotas commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -87,111 +127,80 @@ FCIMPL3(VOID, ExceptionNative::GetStackTracesDeepCopy, Object* pExceptionObjectU
 
-    bool fHaveStackTrace = false;
-    bool fHaveDynamicMethodArray = false;
+    gc.stackTrace.MarkAsFrozen();
```
I think this should happen in `SaveStackTracesFromDeepCopy`.

janvorli committed:
Fix typo

jkotas commented on src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs:
```diff
@@ -147,17 +147,14 @@ internal void RestoreDispatchState(in DispatchState dispatchState)
 
-                // The binary stack trace and references to dynamic methods have to be restored under a lock to guarantee integrity of the system.
-                SaveStackTracesFromDeepCopy(this, stackTraceCopy, dynamicMethodsCopy);
+                SaveStackTracesFromDeepCopy(this, dispatchState.StackTrace);
```
```suggestion
                _stackTrace = dispatchState.StackTrace;
```
And delete `SaveStackTracesFromDeepCopy` fcall?

janvorli replied:
Done

janvorli committed:
Change the size / keepAlive fields in stack trace to uint32_t

Plus a build break fix

janvorli committed:
Remove SaveStackTracesFromDeepCopy

Also rename GetStackTracesDeepCopy to GetFrozenStackTrace and move the
return argument to return value.

jkotas commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -76,75 +76,49 @@ FCIMPL0(VOID, ExceptionNative::PrepareForForeignExceptionRaise)
 
-// Given an exception object, this method will extract the stacktrace and dynamic method array and set them up for return to the caller.
-FCIMPL3(VOID, ExceptionNative::GetStackTracesDeepCopy, Object* pExceptionObjectUnsafe, Object **pStackTraceUnsafe, Object **pDynamicMethodsUnsafe);
+void DeepCopyStackTrace(StackTraceArray &stackTrace, StackTraceArray &stackTraceCopy, PTRARRAYREF *pKeepAliveArray, PTRARRAYREF *pKeepAliveArrayCopy)
```
Delete - no longer used

janvorli replied:
Done

jkotas commented on src/coreclr/vm/clrex.h:
```diff
@@ -51,28 +54,15 @@ struct StackTraceElement
-    unsigned            m_dFrameCount;      // current frame in stack trace
-    unsigned            m_cDynamicMethodItems; // number of items in the Dynamic Method array
-    unsigned            m_dCurrentDynamicIndex; // index of the next location where the resolver object will be stored
+    int m_dummy; // remove this
```
```suggestion
```
?

janvorli replied:
Removed

jkotas commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -157,44 +131,39 @@ FCIMPL3(VOID, ExceptionNative::SaveStackTracesFromDeepCopy, Object* pExceptionOb
-        gc.refException->SetStackTrace(NULL, NULL);
+        gc.result = NULL;
     }
+    
+    // There can be no GC after setting the frozenStackTrace until the Object is returned.
```
```suggestion
```
I am not sure what this comment is trying to say.

janvorli replied:
Ah, I've renamed the frozenStackTrace to result and forgotten to update the comment. It is trying to say that the gc.result is not protected after the HELPER_METHOD_FRAME_END

jkotas replied:
Ok, that's writing manually managed code 101.

(We will want to convert this method to QCALL in near future to get rid of the HELPER_METHOD_FRAME.)

jkotas commented on src/coreclr/vm/comutilnative.cpp:
```diff
@@ -157,44 +131,39 @@ FCIMPL3(VOID, ExceptionNative::SaveStackTracesFromDeepCopy, Object* pExceptionOb
-    // If there is no stacktrace, then there cannot be any dynamic method array. Thus,
-    // save stacktrace only when we have it.
-    if (gc.stackTrace.Size() > 0)
+    else if (gc.stackTrace.Get() != NULL)
```
```suggestion
    else
```
Unnecessary null check

janvorli replied:
REmoved

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3224,198 +2856,316 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
+
+        // Fetch the stacktrace and the keepAlive array from the exception object. It returns clones of those arrays in case the
+        // stack trace was created by a different thread.
+        bool wasCreatedByForeignThread = ((EXCEPTIONREF)ObjectFromHandle(hThrowable))->GetStackTrace(gc.stackTrace, &gc.pKeepAliveArray);
```
```suggestion
        ((EXCEPTIONREF)ObjectFromHandle(hThrowable))->GetStackTrace(gc.stackTrace, &gc.pKeepAliveArray);
```
Unused

janvorli replied:
Removed

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
 #endif // !defined(DACCESS_COMPILE)
 
-void ExceptionObject::GetStackTrace(StackTraceArray & stackTrace, PTRARRAYREF * outDynamicMethodArray /*= NULL*/) const
+bool ExceptionObject::GetStackTrace(StackTraceArray & stackTrace, PTRARRAYREF * outKeepAliveArray /*= NULL*/) const
```
```suggestion
void ExceptionObject::GetStackTrace(StackTraceArray & stackTrace, PTRARRAYREF * outKeepAliveArray /*= NULL*/) const
```

janvorli replied:
Changed

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
-#if !defined(DACCESS_COMPILE)
-    SpinLock::AcquireLock(&g_StackTraceArrayLock);
-#endif // !defined(DACCESS_COMPILE)
+    bool wasCreatedByForeignThread = false;
```
```suggestion
```

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+
+        // When the stack trace was created by other thread than the current one, we create a copy of both the stack trace and the keepAlive arrays to make sure
+        // they are not changing while the caller is accessing them.
+        wasCreatedByForeignThread = true;
```
```suggestion
```

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+        GCPROTECT_END();
+    }
+#endif // DACCESS_COMPILE            
+ 
+    return wasCreatedByForeignThread;
```
```suggestion
 ```

janvorli committed:
Remove dummy field and an unused function

jkotas commented on src/coreclr/vm/clrex.h:
```diff
@@ -51,28 +54,13 @@ struct StackTraceElement
-    BOOL AppendElement(BOOL bAllowAllocMem, UINT_PTR currentIP, UINT_PTR currentSP, MethodDesc* pFunc, CrawlFrame* pCf);
-
-    void GetLeafFrameInfo(StackTraceElement* pStackTraceElement);
+    static BOOL AppendElement(OBJECTHANDLE hThrowable, UINT_PTR currentIP, UINT_PTR currentSP, MethodDesc* pFunc, CrawlFrame* pCf);
```
```suggestion
    static void AppendElement(OBJECTHANDLE hThrowable, UINT_PTR currentIP, UINT_PTR currentSP, MethodDesc* pFunc, CrawlFrame* pCf);
```

janvorli replied:
Changed

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3224,198 +2856,316 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
 
 //
+// Append stack frame to an exception stack trace.
 // Returns true if it appended the element, false otherwise.
```
```suggestion
```

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3224,198 +2856,316 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
 // Returns true if it appended the element, false otherwise.
 //
-BOOL StackTraceInfo::AppendElement(BOOL bAllowAllocMem, UINT_PTR currentIP, UINT_PTR currentSP, MethodDesc* pFunc, CrawlFrame* pCf)
+BOOL StackTraceInfo::AppendElement(OBJECTHANDLE hThrowable, UINT_PTR currentIP, UINT_PTR currentSP, MethodDesc* pFunc, CrawlFrame* pCf)
```
```suggestion
void StackTraceInfo::AppendElement(OBJECTHANDLE hThrowable, UINT_PTR currentIP, UINT_PTR currentSP, MethodDesc* pFunc, CrawlFrame* pCf)
```

jkotas commented on src/coreclr/vm/excep.cpp:
```diff
@@ -3224,198 +2856,316 @@ void COMPlusCooperativeTransitionHandler(Frame* pFrame)
     CONTRACTL_END
 
-    LOG((LF_EH, LL_INFO10000, "StackTraceInfo::AppendElement (%p), IP = %p, SP = %p, %s::%s\n", this, currentIP, currentSP, pFunc ? pFunc->m_pszDebugClassName : "", pFunc ? pFunc->m_pszDebugMethodName : "" ));
     BOOL bRetVal = FALSE;
```
```suggestion
```

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1545,13 +1558,11 @@ void StackTraceArray::CopyFrom(StackTraceArray const & src)
-    m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(src.Capacity()));
-
-    Volatile<size_t> size = src.Size();
+    Volatile<uint32_t> size = src.Size();
```
```suggestion
    uint32_t size = src.Size();
```
Local `Volatile<T>` variable does not make sense. If the Size() reading needs to have read memory barrier, the barrier should be on the reading side.

janvorli replied:
It's been there from the first git commit to the repo. I was trying to reason about whether the memory barrier it introduces is needed too and came to the same conclusion as you.

janvorli replied:
Changed

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+    CONTRACTL
     {
-        *outDynamicMethodArray = _dynamicMethods;
+        GC_TRIGGERS;
+        THROWS;
```
```suggestion
        GC_NOTRIGGER;
        NOTHROW;
```
This does not seem to trigger GC or throw

janvorli replied:
Thanks for spotting this, it used to have an allocation, now it doesn't.

janvorli replied:
removed

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
-#if !defined(DACCESS_COMPILE)
-    SpinLock::ReleaseLock(&g_StackTraceArrayLock);
-#endif // !defined(DACCESS_COMPILE)
+    struct
+    {
+        PTRARRAYREF keepAliveArray = NULL;
+        PTR_PTRArray combinedArray = NULL;
+    } gc;
+
+    GCPROTECT_BEGIN(gc);
```
```suggestion
```
Unnecessary - does not trigger GC

janvorli commented on src/coreclr/vm/object.cpp:
```diff
@@ -1545,13 +1558,11 @@ void StackTraceArray::CopyFrom(StackTraceArray const & src)
-
     SetSize(size);  // set size to the exact value which was used when we copied the data
                     // another thread might have changed it at the time of copying
+    SetKeepAliveItemsCount(ComputeKeepAliveItemsCount());
```
@jkotas I am considering making a change here. If I added a way to get and set the size and keep alive items count atomically together from the stacktrace, I could get rid of the ComputeKeepAliveItemsCount usage for anything else than asserts.

jkotas replied:
I think you will have troubles with alignment on 32-bit platforms to make this work.

janvorli replied:
Hmm, right.

janvorli committed:
Cleanup based on feedback

janvorli commented on src/coreclr/vm/object.h:
```diff
@@ -1914,6 +1914,7 @@ class StackTraceArray
     struct ArrayHeader
     {
         size_t m_size;
+        size_t m_keepAliveItemsCount;
```
Changed

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+            _ASSERTE((*outKeepAliveArray) != NULL);
+            memmoveGCRefs(gc.newKeepAliveArray->GetDataPtr() + 1,
+                          (*outKeepAliveArray)->GetDataPtr() + 1,
+                          (keepAliveArrayCapacity - 1) * sizeof(Object *));
```
We should only copy `j - 1` elements. We do not need to copy the spurious extra elements that were cut off.

(Also, `j` can use better name.)

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1918,130 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+                }
+                if (keepAliveObject == NULL)
+                {
+                    // Trim the stack trace at a point where a dynamic or collectible method is found without a corresponding keepAlive object.
```
The invariants of the lock-free algorithm (and why this is needed) can use better explanation. It is not obvious how we can end up with spurious extra keep alive entries.

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1545,13 +1558,11 @@ void StackTraceArray::CopyFrom(StackTraceArray const & src)
-
     SetSize(size);  // set size to the exact value which was used when we copied the data
                     // another thread might have changed it at the time of copying
+    SetKeepAliveItemsCount(ComputeKeepAliveItemsCount());
```
I do not see why we are recomputing the keep alive items count here.

janvorli replied:
The source stack trace might have new entries added. We've copied just "size" entries, but the keep alive items count may cover some of the entries that were added between the time we've read the size and copied the header and elements. So, we don't really know the actually keep alive count matching the number of elements we've copied, thus we compute it.

jkotas replied:
Is this needed for all callers (there are not that many) or just `ExceptionObject::GetStackTrace`? It seems to be only needed for `ExceptionObject::GetStackTrace`. `ExceptionObject::GetStackTrace` does other consistency checks and we may just fold it into those.

jkotas replied:
I have been trying to identify the operations that are performed on the stacktrace data structure by the rest of the system and review them for thread-safety. It is kind of hard since the operations are not clearly delineated in the code. It would really help if the operations performed on the stacktrace data structure were easy to identify.

 I am thinking about something like:
```
struct StackTrace
{
public:
    // Initialize with OBJECTREF
    StackTrace(OBJECTREF obj);
    // Extract OBJECTREF
    OBJECTREF GetValue();

    // Ensures that the current thread has exclusive write access to the stacktrace and there is space for given number of 
    // extra elements. Creates copy as necessary.
    // (THROWS, GC_TRIGGERS)
    void AcquireExlusiveAccess(int extraElements, int extraKeepAlive);

    // Appends the element to the stacktrace.  Assumes that AcquireExlusiveAccess has been called with extraElements/extraKeepAlive as appropriate.
    // (NOTHROW, GC_NOTRIGGER)
    void AppendStackTraceElement(const StackTraceElement& element, OBJECTREF keepAlive);

    // Freezes the stacktrace. Assumes that AcquireExlusiveAccess has been called.
    // (NOTHROW, GC_NOTRIGGER)
    void Freeze();

    // Get number of valid elements (cuts-off spurious extra elements introduced by race conditions)
    // (NOTHROW, GC_NOTRIGGER)
    int GetNumberOfStackTraceElements();

    // Return i-th stack element. i must be < GetNumberOfStackTraceElements().
    // (NOTHROW, GC_NOTRIGGER)
    StackTraceElement& GetStackTraceAt(int i);

private:
    ... private helper methods as necessary ...
};
```

janvorli replied:
> Is this needed for all callers (there are not that many) or just `ExceptionObject::GetStackTrace`? It seems to be only needed for `ExceptionObject::GetStackTrace`. `ExceptionObject::GetStackTrace` does other consistency checks and we may just fold it into those.

There are just two callers - the `ExceptionObject::GetStackTrace` you've mentioned and the `StackTraceInfo::EnsureStackTraceArray` that always calls that with the stack trace that belongs to the current thread. I could change it so that there are two versions of `CopyDataFrom` / parametrized one where one does just copying of the array and the other called by the `ExceptionObject::GetStackTrace` would do the keep alive count recomputation.

As for the thread safety in general, the only place that needs to take care of stack traces coming from different thread and possibly mutating under our hands is the `ExceptionObject::GetStackTrace`.  So only this method (and the `CopyDataFrom` that's part of it) takes care of the dynamically changing stack trace / keepalive array. It always returns a stack trace that is owned by the current thread, so it can be safely iterated / modified. This is the key point of the new design. 

The only special case when the copy doesn't happen is in DAC.

jkotas replied:
It does not look right that `CopyDataFrom` compensates for some inconsistencies introduced by race conditions, but not all of them. And that this compensation is actually only needed by one of the callers.

I think it would be best to get rid of the compensation for race conditions from `CopyDataFrom` and inline it into the `ExceptionObject::GetStackTrace` instead. It can be also more efficient that way. The existing loop in `ExceptionObject::GetStackTrace` can be used to prune both spurious StackTraceElements and spurious keepAlives.

janvorli replied:
Done

janvorli committed:
Move the race handling into GetStackTrace only

Plus an unused method removal and a little naming / contract cleanup

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1906,40 +1898,131 @@ void ExceptionObject::SetStackTrace(I1ARRAYREF stackTrace, PTRARRAYREF dynamicMe
+        uint32_t keepAliveArrayCapacity = ((*outKeepAliveArray) == NULL) ? 0 : (*outKeepAliveArray)->GetNumComponents();
+
+        uint32_t keepAliveItemsCount = 0;
+        uint32_t count = stackTrace.Size();
```
```suggestion
```

jkotas commented on src/coreclr/vm/object.cpp:
```diff
@@ -1464,95 +1464,86 @@ void StackTraceArray::CheckState() const
-    m_array = (I1ARRAYREF) AllocatePrimitiveArray(ELEMENT_TYPE_I1, static_cast<DWORD>(src.Capacity()));
-
-    Volatile<size_t> size = src.Size();
+    uint32_t size = src.Size();
```
We should use volatile read for the Size. Otherwise, the C++ compiler is free to duplicate the read and use one `size` for value for the memcpy and return a different value from the method.

janvorli replied:
I have thought you were suggesting before that I get rid of the Volatile here, so I am bit confused.

jkotas replied:
I would:
- Get rid of the volatile local variable. It does not make sense to tell the compiler to issue read/write barriers when reading/writing stack. It is deoptimization.
- Change the read in Size() method to be volatile read, so that it is not reordered. This volatile read is counterpart of volatile write in `StackTraceArray::Append` (it is not actual volatile write, but MemoryBarrier + regular write that is equivalent in this situation). In general, volatile reads and volatile writes in lock-free algorithm have to come in pairs on producing/consuming sides.

janvorli replied:
Ok, I have added a commit with that change. I've modified the GetSize / SetSize and added such treatment to the keep alive count accessors too. And removed the MemoryBarrier from the StackTraceArray::Append where it is no longer needed after this change.

jkotas reviewed (approved):
LGTM otherwise. Thank you!

jkotas commented on src/coreclr/inc/sospriv.idl:
```diff
@@ -444,7 +444,7 @@ interface ISOSDacInterface8 : IUnknown
 // stress log structs (StressMsg, StressLogChunck, ThreadStressLog, etc), exception
 // stack traces (StackTraceElement), the PredefinedTlsSlots enums, etc.
-cpp_quote("#define SOS_BREAKING_CHANGE_VERSION 4")
+cpp_quote("#define SOS_BREAKING_CHANGE_VERSION 5")
```
cc @mikem8361 `SOS_BREAKING_CHANGE_VERSION` bump

mikem8361 replied:
This is good, but we need to bump the version to 5 in the diagnostics repo.  We should have done that when Jan updated SOS for the Exception's _stackTrace changes.

janvorli replied:
Ah, I didn't know there is a counterpart for this in the diag repo.

mikem8361 replied:
I should have caught this in the diag PR. I have a PR up to make this change and hopefully get into our pending tools release.

janvorli committed:
Add VolatileLoad/Store around the size / keep alive count

Also remove the memory barrier from the StackTraceArray::Append since it
is not needed after that change.

janvorli commented:
This PR needs to wait for merging until SOS with corresponding change is released.

janvorli committed:
Add comment on why trimming the stack trace by keep alive is needed

I have also realized that when we need to trim, the keepAlive array is
always fully populated, so we don't need to check for cases where there
would be NULL in an entry of the array.

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

janvorli commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc
